File: mobile/android/src/main/java/com/amigo/android/auth/AuthViewModel.kt
package com.amigo.android.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.amigo.shared.auth.EmailAuthenticator
import com.amigo.shared.auth.OAuthAuthenticator
import com.amigo.shared.auth.SessionManager
import com.amigo.shared.auth.models.AuthResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AuthViewModel(
    private val emailAuthenticator: EmailAuthenticator,
    private val oauthAuthenticator: OAuthAuthenticator,
    private val sessionManager: SessionManager
) : ViewModel() {
    
    private val _email = MutableStateFlow("")
    val email: StateFlow<String> = _email.asStateFlow()
    
    private val _password = MutableStateFlow("")
    val password: StateFlow<String> = _password.asStateFlow()
    
    private val _confirmPassword = MutableStateFlow("")
    val confirmPassword: StateFlow<String> = _confirmPassword.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    private val _successMessage = MutableStateFlow<String?>(null)
    val successMessage: StateFlow<String?> = _successMessage.asStateFlow()
    
    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()
    
    init {
        viewModelScope.launch {
            sessionManager.initialize()
            sessionManager.isAuthenticated.collect { authenticated ->
                _isAuthenticated.value = authenticated
            }
        }
    }
    
    fun updateEmail(value: String) {
        _email.value = value
    }
    
    fun updatePassword(value: String) {
        _password.value = value
    }
    
    fun updateConfirmPassword(value: String) {
        _confirmPassword.value = value
    }
    
    fun signIn() {
        if (_email.value.isEmpty() || _password.value.isEmpty()) {
            _errorMessage.value = "Please enter email and password"
            return
        }
        
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                val result = emailAuthenticator.signIn(_email.value, _password.value)
                
                when (result) {
                    is AuthResult.Success -> {
                        sessionManager.saveSession(result.session)
                    }
                    is AuthResult.EmailConfirmationRequired -> {
                        // Email confirmation not needed for sign in
                    }
                    is AuthResult.Error -> {
                        // Provide user-friendly message for email not confirmed
                        _errorMessage.value = when (result.code) {
                            com.amigo.shared.auth.models.AuthErrorCode.EMAIL_NOT_CONFIRMED -> 
                                "Please confirm your email address before signing in. Check your inbox for the confirmation link."
                            else -> result.message
                        }
                    }
                }
            } catch (e: Exception) {
                _errorMessage.value = "Sign in failed: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun signUp() {
        if (_email.value.isEmpty() || _password.value.isEmpty()) {
            _errorMessage.value = "Please enter email and password"
            return
        }
        
        if (_password.value != _confirmPassword.value) {
            _errorMessage.value = "Passwords do not match"
            return
        }
        
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            _successMessage.value = null
            
            try {
                val result = emailAuthenticator.signUp(_email.value, _password.value)
                
                when (result) {
                    is AuthResult.Success -> {
                        sessionManager.saveSession(result.session)
                    }
                    is AuthResult.EmailConfirmationRequired -> {
                        _successMessage.value = "Account created! Please check your email (${result.email}) to confirm your account."
                    }
                    is AuthResult.Error -> {
                        _errorMessage.value = result.message
                    }
                }
            } catch (e: Exception) {
                _errorMessage.value = "Sign up failed: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun signInWithGoogle() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                val result = oauthAuthenticator.signInWithGoogle()
                
                when (result) {
                    is AuthResult.Success -> {
                        sessionManager.saveSession(result.session)
                    }
                    is AuthResult.EmailConfirmationRequired -> {
                        // Email confirmation not needed for OAuth
                    }
                    is AuthResult.Error -> {
                        _errorMessage.value = result.message
                    }
                }
            } catch (e: Exception) {
                _errorMessage.value = "Google sign in failed: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun signOut() {
        viewModelScope.launch {
            try {
                sessionManager.signOut()
                _email.value = ""
                _password.value = ""
                _confirmPassword.value = ""
            } catch (e: Exception) {
                _errorMessage.value = "Sign out failed: ${e.message}"
            }
        }
    }
    
    fun clearError() {
        _errorMessage.value = null
    }
    
    fun clearSuccess() {
        _successMessage.value = null
    }
    
    suspend fun handleDeepLinkSession(accessToken: String, refreshToken: String) {
        _isLoading.value = true
        _errorMessage.value = null
        
        try {
            val result = sessionManager.handleDeepLinkSession(accessToken, refreshToken)
            
            when (result) {
                is AuthResult.Success -> {
                    // Session imported successfully
                    _isAuthenticated.value = true
                }
                is AuthResult.EmailConfirmationRequired -> {
                    // Not applicable for deep link
                }
                is AuthResult.Error -> {
                    _errorMessage.value = result.message
                }
            }
        } catch (e: Exception) {
            _errorMessage.value = "Failed to handle session: ${e.message}"
        } finally {
            _isLoading.value = false
        }
    }
}

File: mobile/ios/Amigo/Auth/AuthViewModel.swift
import SwiftUI
import shared
import Combine

@MainActor
class AuthViewModel: ObservableObject {
    @Published var email: String = ""
    @Published var password: String = ""
    @Published var confirmPassword: String = ""
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    @Published var successMessage: String?
    @Published var isAuthenticated: Bool = false
    
    private let emailAuthenticator: EmailAuthenticator
    private let oauthAuthenticator: OAuthAuthenticator
    internal let sessionManager: SessionManager
    
    init(emailAuthenticator: EmailAuthenticator, oauthAuthenticator: OAuthAuthenticator, sessionManager: SessionManager) {
        self.emailAuthenticator = emailAuthenticator
        self.oauthAuthenticator = oauthAuthenticator
        self.sessionManager = sessionManager
        
        // Initialize session manager
        Task {
            await initializeSession()
        }
    }
    
    func initializeSession() async {
        print("🔵 [iOS AuthViewModel] initializeSession() START")
        do {
            print("🔵 [iOS AuthViewModel] Calling sessionManager.initialize()...")
            try await sessionManager.initialize()
            let authStatus = sessionManager.isAuthenticated.value as! Bool
            print("🔵 [iOS AuthViewModel] SessionManager initialized, isAuthenticated=\(authStatus)")
            isAuthenticated = authStatus
            
            if authStatus {
                let user = try await sessionManager.getCurrentUser()
                print("✅ [iOS AuthViewModel] User logged in: id=\(user?.id ?? "nil"), email=\(user?.email ?? "nil")")
            } else {
                print("⚠️ [iOS AuthViewModel] No authenticated user")
            }
        } catch {
            print("❌ [iOS AuthViewModel] Failed to initialize session: \(error)")
        }
    }
    
    func signIn() async {
        guard !email.isEmpty && !password.isEmpty else {
            errorMessage = "Please enter email and password"
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        do {
            let result = try await emailAuthenticator.signIn(email: email, password: password)
            
            if let success = result as? AuthResult.Success {
                // SDK 3.x automatically saves the session
                isAuthenticated = true
            } else if let error = result as? AuthResult.Error {
                // Provide user-friendly message for email not confirmed
                if error.code == .emailNotConfirmed {
                    errorMessage = "Please confirm your email address before signing in. Check your inbox for the confirmation link."
                } else {
                    errorMessage = error.message
                }
            }
        } catch {
            errorMessage = "Sign in failed: \(error.localizedDescription)"
        }
        
        isLoading = false
    }
    
    func signUp() async {
        guard !email.isEmpty && !password.isEmpty else {
            errorMessage = "Please enter email and password"
            return
        }
        
        guard password == confirmPassword else {
            errorMessage = "Passwords do not match"
            return
        }
        
        isLoading = true
        errorMessage = nil
        successMessage = nil
        
        do {
            let result = try await emailAuthenticator.signUp(email: email, password: password)
            
            if let success = result as? AuthResult.Success {
                // SDK 3.x automatically saves the session
                isAuthenticated = true
            } else if let confirmation = result as? AuthResult.EmailConfirmationRequired {
                successMessage = "Account created! Please check your email (\(confirmation.email)) to confirm your account."
            } else if let error = result as? AuthResult.Error {
                errorMessage = error.message
            }
        } catch {
            errorMessage = "Sign up failed: \(error.localizedDescription)"
        }
        
        isLoading = false
    }
    
    func signInWithGoogle() async {
        print("🔐 [iOS AuthViewModel] signInWithGoogle() START")
        isLoading = true
        errorMessage = nil
        
        do {
            print("🔐 [iOS AuthViewModel] Calling oauthAuthenticator.signInWithGoogle()...")
            let result = try await oauthAuthenticator.signInWithGoogle()
            
            if let success = result as? AuthResult.Success {
                print("✅ [iOS AuthViewModel] Google sign-in SUCCESS")
                print("✅ [iOS AuthViewModel] Session user.id: \(success.session.user.id)")
                print("✅ [iOS AuthViewModel] Session email: \(success.session.user.email)")
                print("✅ [iOS AuthViewModel] Access token length: \(success.session.accessToken.count)")
                print("✅ [iOS AuthViewModel] SDK 3.x automatically saves session")
                isAuthenticated = true
            } else if let error = result as? AuthResult.Error {
                print("❌ [iOS AuthViewModel] Google sign-in ERROR: \(error.message)")
                errorMessage = error.message
            }
        } catch {
            print("❌ [iOS AuthViewModel] Google sign-in EXCEPTION: \(error.localizedDescription)")
            errorMessage = "Google sign in failed: \(error.localizedDescription)"
        }
        
        isLoading = false
        print("🔐 [iOS AuthViewModel] signInWithGoogle() END")
    }
    
    func signInWithApple() async {
        print("🔐 [iOS AuthViewModel] signInWithApple() START")
        isLoading = true
        errorMessage = nil
        
        do {
            print("🔐 [iOS AuthViewModel] Calling oauthAuthenticator.signInWithApple()...")
            let result = try await oauthAuthenticator.signInWithApple()
            
            if let success = result as? AuthResult.Success {
                print("✅ [iOS AuthViewModel] Apple sign-in SUCCESS")
                print("✅ [iOS AuthViewModel] Session user.id: \(success.session.user.id)")
                print("✅ [iOS AuthViewModel] Session email: \(success.session.user.email)")
                print("✅ [iOS AuthViewModel] Access token length: \(success.session.accessToken.count)")
                print("✅ [iOS AuthViewModel] SDK 3.x automatically saves session")
                isAuthenticated = true
            } else if let error = result as? AuthResult.Error {
                print("❌ [iOS AuthViewModel] Apple sign-in ERROR: \(error.message)")
                errorMessage = error.message
            }
        } catch {
            print("❌ [iOS AuthViewModel] Apple sign-in EXCEPTION: \(error.localizedDescription)")
            errorMessage = "Apple sign in failed: \(error.localizedDescription)"
        }
        
        isLoading = false
        print("🔐 [iOS AuthViewModel] signInWithApple() END")
    }
    
    func signOut() async {
        do {
            try await sessionManager.signOut()
            isAuthenticated = false
            email = ""
            password = ""
            confirmPassword = ""
        } catch {
            errorMessage = "Sign out failed: \(error.localizedDescription)"
        }
    }
    
    func clearError() {
        errorMessage = nil
    }
    
    func clearSuccess() {
        successMessage = nil
    }
    
    func getCurrentUser() async throws -> User? {
        return try await sessionManager.getCurrentUser()
    }
    
    func handleDeepLinkSession(accessToken: String, refreshToken: String, expiresIn: Int64 = 3600) async {
        print("🔗 [iOS AuthViewModel] handleDeepLinkSession() START")
        print("🔗 [iOS AuthViewModel] Access token length: \(accessToken.count)")
        print("🔗 [iOS AuthViewModel] Refresh token length: \(refreshToken.count)")
        print("🔗 [iOS AuthViewModel] Expires in: \(expiresIn) seconds")
        isLoading = true
        errorMessage = nil
        
        do {
            print("🔗 [iOS AuthViewModel] Setting session from tokens...")
            try await sessionManager.setSessionFromTokens(accessToken: accessToken, refreshToken: refreshToken, expiresIn: expiresIn)
            
            print("✅ [iOS AuthViewModel] Session set successfully")
            // Trust that the session is valid - don't check for user
            // The user field may be null initially but the session is authenticated
            isAuthenticated = true
            print("✅ [iOS AuthViewModel] User authenticated via deep link")
        } catch {
            print("❌ [iOS AuthViewModel] Deep link session EXCEPTION: \(error.localizedDescription)")
            errorMessage = "Failed to handle session: \(error.localizedDescription)"
        }
        
        isLoading = false
        print("🔗 [iOS AuthViewModel] handleDeepLinkSession() END")
    }
}
